analysis/quantifier-comparisons/15-demux-comparisons.Rmd

```diff
@@ -0,0 +1,290 @@
+  library(SingleCellExperiment)
+})
+
+set.seed(2021)
+```
+
+File paths:
+
+```{r}
+file_dir <- file.path("data", "multiplex")
+
+
+# singleCellExperiment files from 
+filtered_file <- file.path(file_dir, paste0(params$lib_id, "_filtered.rds"))
+
+
+## outfiles
+outdir <- "results"
+dir.create(outdir, recursive = TRUE, showWarnings = FALSE)
+demux_tsv <- file.path(outdir, paste0(params$lib_id, "_demux.tsv"))
+demux_tables <- file.path(outdir, paste0(params$lib_id, "_demux_table.txt"))
+
+plotdir <- "plots"
+demux_umap_plot <- file.path(plotdir, paste0(params$lib_id, "_demux_umap.png"))
+```
+
+## Functions
+
+```{r}
+# calculate the agreement between two lists, with and or without NAs included
+calc_agreement <- function(call_a, call_b, na.include = FALSE){
+  agree <- sum(call_a == call_b, na.rm = TRUE)
+  disagree <- sum(call_a != call_b, na.rm = TRUE)
+  if(na.include){
+    # add single NA values
+    disagree <- disagree  + sum(xor(is.na(call_a), is.na(call_b)))
+  } 
+  return(agree/(agree + disagree))
+}
+
+```
+
+
+
+
+# Single cell processing and filtering
+
+The SingleCellExperiment files have already been processed via the `scpca-nf` pipeline, which performs filtering and adds cellhash-based demultiplexing calls using `DropletUtils::HashedDrops` and `Seurat::HTODemux`.
+
+The pipeline also incorporates genetic information inferred from bulk sequencing and performs genetic demultiplexing with [cellsnp](https://cellsnp-lite.readthedocs.io/en/latest/) and [Vireo](https://vireosnp.readthedocs.io/en/latest/).
+
+
+```{r}
+sce <- readRDS(filtered_file)
+```
+
+## Sample information
+
+The metadata in the processed SingleCellExperiment includes information about the samples included in this library:
+
+```{r}
+metadata(sce)$sample_id
+```
+
+
+# Demultiplexing results
+
+The confident calls (by default measures) are stored in the `colData` data frame, prefixed by method, so we will extract those calls for comparisons.
+Doublets here are coded as part of the `NA` group.
+
+```{r}
+demux_calls <- data.frame(
+  hashedDrops = sce$hashedDrops_sampleid,
+  HTODemux = sce$HTODemux_sampleid,
+  vireo = sce$vireo_sampleid
+)
+
+# Summarize all of the calls
+demux_calls |>
+  dplyr::mutate_all(as.factor) |>
+  summary()
+```
+
+In general, we see that `DropletUtils::hashedDrops` tends to make many fewer calls than `Seurat::HTODemux`. 
+In either case, we have a large excess of `NA` calls, where the cellhash-based demultiplexing algorithms did not find a single best call. 
+
+This may partly be a function of the quality of the cellhash results for this particular library.
+As seen in `14-cellhash-exploration.Rmd`, the cellhash data for these samples seems to have a lot of background signal that may make accurate demultiplexing based on cellhash data alone difficult.
+
+The genetic demultiplexing calls made by `vireo` seem to identify the most cells, but the relative frequencies of the samples do not necessarily agree with the results from `HTODemux`.
```

Comment:
You may have a better understanding of this than I do since I can only see the notebook generated for this single sample, but for SCPCL000535, `vireo` actually doesn't have the most calls. It has similar calls to Seurat, but technically Seurat has the lowest number of `NA`. Is that the case across the board where the total number of `NA` seem to be more similar between `vireo` and Seurat and then `DropletUtils` is lower? If that's the case I think I would change this comment to make the claim that these two methods are more similar in total calls, but the frequencies of the samples don't always agree. Or maybe say that in the majority of cases, ` vireo` has more calls than Seurat if that's the observation you see?

Reply:
Oh, I think the one I uploaded was 537, actually, but I meant it to be 535. But yeah, I can/should rephrase to be a bit more general. (Also adding a bit of code to write out the library name).